chain/consensus/common/crossmsg.go

```diff
@@ -0,0 +1,180 @@
+	}
+
+	// NOTE: Any additional check required?
+	// TODO: ApplyMsg to see that it is correct?
+	return nil
+
+}
+
+func ApplyCrossMsg(ctx context.Context, vmi *vm.VM, submgr subnet.SubnetMgr,
+	em stmgr.ExecMonitor, msg *types.Message,
+	ts *types.TipSet, netName dtypes.NetworkName) error {
+	switch hierarchical.GetMsgType(msg) {
+	case hierarchical.Fund:
+		return applyFundMsg(ctx, vmi, submgr, em, msg, ts, netName)
+	case hierarchical.Release:
+		panic("Not implemented")
+	case hierarchical.Cross:
+		panic("Not implemented")
+	}
+
+	return xerrors.Errorf("Unknown cross-msg type")
+}
+
+func applyFundMsg(ctx context.Context, vmi *vm.VM, submgr subnet.SubnetMgr,
+	em stmgr.ExecMonitor, msg *types.Message, ts *types.TipSet,
+	netName dtypes.NetworkName) error {
+	// sanity-check: the root chain doesn't support topDown messages,
+	// so return an error if submgr is nil.
+	if submgr == nil {
+		return xerrors.Errorf("Root chain doesn't have parent and doesn't support topDown cross msgs")
+	}
+
+	// Get SECPK address for ID from parent chain included in message.
+	id := hierarchical.SubnetID(netName)
+	api, err := submgr.GetSubnetAPI(id.Parent())
+	if err != nil {
+		return xerrors.Errorf("getting subnet API: %w", err)
+	}
+	secpaddr, err := api.StateAccountKey(ctx, msg.From, types.EmptyTSK)
+	if err != nil {
+		return xerrors.Errorf("getting secp address: %w", err)
+	}
+	// Translating parent actor ID of address to SECPK for its application
+	// in subnet.
+	msg.From = secpaddr
+	msg.To = secpaddr
+	return applyMsg(ctx, vmi, em, msg, ts)
+}
+
+func applyMsg(ctx context.Context, vmi *vm.VM, em stmgr.ExecMonitor,
+	msg *types.Message, ts *types.TipSet) error {
+	// Serialize params
+	params := &sca.ApplyParams{
+		Msg: *msg,
+	}
+	serparams, err := actors.SerializeParams(params)
+	if err != nil {
+		return xerrors.Errorf("failed serializing init actor params: %s", err)
+	}
+	apply := &types.Message{
+		From:       builtin.SystemActorAddr,
+		To:         hierarchical.SubnetCoordActorAddr,
+		Nonce:      msg.Nonce,
+		Value:      big.Zero(),
+		GasFeeCap:  types.NewInt(0),
+		GasPremium: types.NewInt(0),
+		GasLimit:   1 << 30,
+		Method:     sca.Methods.ApplyMessage,
+		Params:     serparams,
+	}
+	ret, actErr := vmi.ApplyImplicitMessage(ctx, apply)
+	if actErr != nil {
+		return xerrors.Errorf("failed to apply cross message :%w", actErr)
+	}
+	if em != nil {
+		if err := em.MessageApplied(ctx, ts, apply.Cid(), apply, ret, true); err != nil {
+			return xerrors.Errorf("callback failed on reward message: %w", err)
+		}
+	}
+
+	if ret.ExitCode != 0 {
+		return xerrors.Errorf("reward application message failed (exit %d): %s", ret.ExitCode, ret.ActorErr)
+	}
+	fmt.Println(">>>>>>>>>>> IMPLICIT FUND MESSAGE APPLIED", apply.Cid(), ret)
+	return nil
+}
+
+func getSCAState(ctx context.Context, sm *stmgr.StateManager, submgr subnet.SubnetMgr, id hierarchical.SubnetID) (*sca.SCAState, blockadt.Store, error) {
```

Comment:
Does this need checking? I don't get much 😅

Reply:
This is just picking up the current state in the SCA of the subnet. IIUC there is no checking going on here.